src/test/rbd_mirror/test_mock_PoolWatcher.cc

```diff
@@ -0,0 +1,809 @@
+    return MockMirroringWatcher::get_instance().get_oid();
+  }
+};
+
+MockMirroringWatcher *MockMirroringWatcher::s_instance = nullptr;
+MirroringWatcher<MockTestImageCtx> *MirroringWatcher<MockTestImageCtx>::s_instance = nullptr;
+
+} // namespace librbd
+
+namespace rbd {
+namespace mirror {
+
+template <>
+struct Threads<librbd::MockTestImageCtx> {
+  MockSafeTimer *timer;
+  Mutex &timer_lock;
+
+  MockContextWQ *work_queue;
+
+  Threads(Threads<librbd::ImageCtx> *threads)
+    : timer(new MockSafeTimer()),
+      timer_lock(threads->timer_lock),
+      work_queue(new MockContextWQ()) {
+  }
+  ~Threads() {
+    delete timer;
+    delete work_queue;
+  }
+};
+
+namespace pool_watcher {
+
+template <>
+struct RefreshImagesRequest<librbd::MockTestImageCtx> {
+  ImageIds *image_ids = nullptr;
+  Context *on_finish = nullptr;
+  static RefreshImagesRequest *s_instance;
+  static RefreshImagesRequest *create(librados::IoCtx &io_ctx,
+                                      ImageIds *image_ids,
+                                      Context *on_finish) {
+    assert(s_instance != nullptr);
+    s_instance->image_ids = image_ids;
+    s_instance->on_finish = on_finish;
+    return s_instance;
+  }
+
+  MOCK_METHOD0(send, void());
+
+  RefreshImagesRequest() {
+    s_instance = this;
+  }
+};
+
+RefreshImagesRequest<librbd::MockTestImageCtx> *RefreshImagesRequest<librbd::MockTestImageCtx>::s_instance = nullptr;
+
+} // namespace pool_watcher
+
+} // namespace mirror
+} // namespace rbd
+
+// template definitions
+#include "tools/rbd_mirror/PoolWatcher.cc"
+
+namespace rbd {
+namespace mirror {
+
+using ::testing::_;
+using ::testing::DoAll;
+using ::testing::InSequence;
+using ::testing::Invoke;
+using ::testing::Return;
+using ::testing::StrEq;
+using ::testing::WithArg;
+using ::testing::WithoutArgs;
+
+class TestMockPoolWatcher : public TestMockFixture {
+public:
+  typedef PoolWatcher<librbd::MockTestImageCtx> MockPoolWatcher;
+  typedef Threads<librbd::MockTestImageCtx> MockThreads;
+  typedef pool_watcher::RefreshImagesRequest<librbd::MockTestImageCtx> MockRefreshImagesRequest;
+  typedef librbd::MockMirroringWatcher MockMirroringWatcher;
+  typedef librbd::MirroringWatcher<librbd::MockTestImageCtx> MirroringWatcher;
+
+  struct MockListener : MockPoolWatcher::Listener {
+    TestMockPoolWatcher *test;
+
+    MockListener(TestMockPoolWatcher *test) : test(test) {
+    }
+
+    MOCK_METHOD2(handle_update, void(const ImageIds &, const ImageIds &));
+  };
+
+  TestMockPoolWatcher() : m_lock("TestMockPoolWatcher::m_lock") {
+  }
+
+  void expect_work_queue(MockThreads &mock_threads) {
+    EXPECT_CALL(*mock_threads.work_queue, queue(_, _))
+      .WillRepeatedly(Invoke([this](Context *ctx, int r) {
+          m_threads->work_queue->queue(ctx, r);
+        }));
+  }
+
+  void expect_mirroring_watcher_is_unregister(MockMirroringWatcher &mock_mirroring_watcher,
+                                              bool unregistered) {
```

Comment:
@dillaman May be the function name was supposed to be  `expect_mirroring_watcher_is_unregistered`?